salt/utils/nb_popen.py

```diff
@@ -37,7 +38,37 @@
 log = logging.getLogger(__name__)
 
 
-class NonBlockingPopen(subprocess.Popen):
+class FdPopen(subprocess.Popen):
+    '''
+    Python2 closing file descriptors in a blind way by a system's given range.
+    Therefore close_fds can significantly slow down the system in various circumstances.
+    '''
+
+    if sys.platform.startswith('linux') and salt.ext.six.PY2:
+        def _close_fds(self, but):
+            '''
+            Close only file descriptors that are needed.
+            The issue is not new: https://bugs.python.org/issue1663329
+
+            :param but:
+            :return:
+            '''
+            maxfds = len(os.listdir('/proc/{}/fd'.format(os.getpid())))
```

Comment:
This isn't actually safe; the number of fds open isn't guaranteed to equate to the highest FD open.
This should instead be:
`maxfds = max(int(x) for x in os.listdir('/proc/{}/fd'.format(os.getpid()))` with an appropriate check for if there is somehow no FDS open which would result in an exception from `max([])`